src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/AbstractDocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,232 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixes
+{
+    /// <summary>
+    /// Provides a base class to write a <see cref="FixAllProvider"/> that fixes documents independently.
+    /// </summary>
+    internal abstract class AbstractDocumentBasedFixAllProvider : FixAllProvider
```

Comment:
this is now the base type of DocumentBasedFixAllProvider and SyntaxEditorBasedFixAllProvider.  It is responsible for hte overall strategy of fix-all, while the subclasses are responsible for getting the changed doc contents.

Reply:
Note: the benefits of using this types are substantial.  With the normal batch fixer, we have to do things much less efficiently (as we have no idea which fixes may touch any particular file).  that means we have to run *everything*, and hold everything in memory, and then hope we haven't OOMed by the time we apply things.

We should strongly consider if we should just make htis public and doc the heck out of how to use it properly (you just need to supply two pretty simple overrides).  We should try to push as much of our own external project fixes to then just use this, as well as evangelizing this to the community as the appropriate way to do batch fixing *if* your diagnostics/code-actions work just by fixing the same document they were reported in.

We could also do something similar for diagnostics that touch other files *but* which know they will only fixup the project the diagnostic was reported in.  We'd then get the same benefits from this type.  The only poorly behaving cases would then be ones where the fixer truly did fixup the entire solution and wanted batching.

@mavasani

Reply:
> We should strongly consider if we should just make htis public and doc the heck out of how to use it properly

That would be great. Basically, the user story would be:

1. I want the most simple FixAll implementation: Use static BatchFixer instance
2. I would prefer the FixAll implementation with best performance, at expense of some additional implementation complexity: Sub-type and implement `AbstractDocumentBasedFixAllProvider`
3. I have a very custom FixAll requirements: Sub-type and implement `FixAllProvider`